### PR TITLE
Tf documentation

### PR DESCRIPTION
**Pull request by ethantang95**

### CODE REVIEW

**Comment by ethantang95** [comment]
@timzaman @gheinrich for review, or should we have a meeting and discuss about this? Thanks

File: docs/GettingStartedTensorflow.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
+# Getting Started with Tensorflow 1.1 in DIGITS
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: remove 1.1

File: docs/BuildTensorflow.md [hunk 2]
```diff
@@ -1,19 +1,44 @@
-# Building Torch
+# Installing Tensorflow
```
Comment (TimZaman) [inline docs/BuildTensorflow.md]: Tensorflow->TensorFlow and add ™ when it makes sense.

File: docs/BuildTensorflow.md [hunk 3]
```diff
@@ -1,19 +1,44 @@
...
+Installation for [Ubuntu](https://www.tensorflow.org/install/install_linux#installing_with_virtualenv)
 
-We recommend installing Tensorflow in a fixed and separate environment.
+Installation for [Mac](https://www.tensorflow.org/install/install_mac#installing_with_virtualenv)
 
-Installation per: https://www.tensorflow.org/versions/r0.10/get_started/os_setup.html#virtualenv-installation
+## Requirements
 
-## Prerequisites
+DIGITS is current targetting Tensorflow-gpu V1.1.
```
Comment (TimZaman) [inline docs/BuildTensorflow.md]: "Tensorflow-gpu" to "Tensorflow 1.1.0rc0"

File: docs/BuildTensorflow.md [hunk 4]
```diff
@@ -1,19 +1,44 @@
-# Building Torch
+# Installing Tensorflow
 
 DIGITS now supports Tensorflow as an optional alternative backend to Caffe or Torch.
 
 > NOTE: Tensorflow support is still experimental!
```
Comment (TimZaman) [inline docs/BuildTensorflow.md]: remove

File: docs/BuildTensorflow.md [hunk 5]
```diff
@@ -1,19 +1,44 @@
-# Building Torch
+# Installing Tensorflow
 
 DIGITS now supports Tensorflow as an optional alternative backend to Caffe or Torch.
```
Comment (TimZaman) [inline docs/BuildTensorflow.md]: remove

File: docs/GettingStartedTensorflow.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+DIGITS will automatically enable support for Tensorflow if it detects that Tensorflow-gpu is installed in the system. This is done by a line of python code that attempts to ```import tensorflow``` to see if it actually imports.
+
+If DIGITS cannot enable tensorflow, a message will be printed in the console saying: ```Tensorflow support is disabled```
+
+## Selecting Tensorflow When Creating A Model In DIGITS
+
+Click on the "Tensorflow" tab on the model creation page
+
+{insert image here}
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: add

File: docs/GettingStartedTensorflow.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+## Defining A Tensorflow Model In DIGITS
+
+To define a Tensorflow model in DIGITS, you need to write a python class that follows this basic template
+
+```python
+class UserModel(Tower):
+*
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: star?

File: docs/GettingStartedTensorflow.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+{insert image here}
+
+## Defining A Tensorflow Model In DIGITS
+
+To define a Tensorflow model in DIGITS, you need to write a python class that follows this basic template
+
+```python
+class UserModel(Tower):
+* 
+    @model_propertyOther Tensorflow Tools in DIGITS
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: ~Other Tensorflow *~

File: docs/GettingStartedTensorflow.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        # Your code here
+        return model
+
+    @model_property
+    def loss(self):
+        # Your code here
+        return loss
+```
+
+For example, this is what it looks like for LeNet:
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: for a LeNet-5 model, with LeNet-5 clickable to http://yann.lecun.com/exdb/lenet/

File: docs/GettingStartedTensorflow.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+The properties ```inference``` and ```loss``` must be defined and the class must be called ```UserModel``` and it must inherit ```Tower```. This is how DIGITS will interact with the python code.
+
+### Provided Properties
+
+Properties that are accessible through ```self```
+
+Property name | Type      | Description
+--------------|-----------|------------
+nclasses      | number    | Number of classes (for classification datasets). For other type of datasets, this is undefined
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: nclasses -> `nclasses`
and do for the rest of the table properties too

File: docs/GettingStartedTensorflow.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+x             | Tensor    | The input node, with the shape of [N, H, W, C]
+y             | Tensor    | The label, [N] for scalar labels, [N, H, W, C] otherwise. Defined only if self.is_training is True
+
+### Internal Properties
+
+These properties are in the ```UserModel``` class written by the user
+
+Property name | Return Type | Description
+--------------|-------------|------------
+__init()__    | None        | The constructor for the ```UserModel``` class
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: double dunders do not render! also use backticks

Comment (TimZaman) [reply docs/GettingStartedTensorflow.md]: init-> `__init__`

Comment (TimZaman) [reply docs/GettingStartedTensorflow.md]: also for the other methods (!)

File: docs/GettingStartedTensorflow.md [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+Property name | Type      | Description
+--------------|-----------|------------
+nclasses      | number    | Number of classes (for classification datasets). For other type of datasets, this is undefined
+input_shape   | Tensor    | Shape (1D Tensor) of the first input Tensor. For image data, this is set to height, width, and channels accessible by [0], [1], and [2] respectively.
+is_training   | boolean   | Whether this is a training graph
+is_inference  | boolean   | Whether this graph is created for inference/testing
+x             | Tensor    | The input node, with the shape of [N, H, W, C]
+y             | Tensor    | The label, [N] for scalar labels, [N, H, W, C] otherwise. Defined only if self.is_training is True
+
+### Internal Properties
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: Properties -> Methods, everywhere

File: docs/GettingStartedTensorflow.md [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+## Other Tensorflow Tools in DIGITS
+
+DIGITS provides a few useful tools to help with your development with Tensorflow.
+
+### Provided Helpful Functions
+
+DIGITS provides a few helpful functions to help you with creating the model. Here are the functions we provide
+
+Function Name       | Parameters          | Description
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: All function names should have backtics

Comment (TimZaman) [reply docs/GettingStartedTensorflow.md]: Also add backtics to parametrs like:
pred - the images (..) -> `pred` - the images (..)

File: docs/GettingStartedTensorflow.md [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+hwc_to_chw | x - the tensor to transpose | Transpose the tensor that was originally HWC format to CHW. The tensor must be a degree of 3
+chw_to_hwc | x - the tensor to transpose | Transpose the tensor that was originally CHW format to HWC. The tensor must be a degree of 3
+bgr_to_rgb | x - the tensor to transform | Transform the tensor that was originally in BGR channels to RGB.
+rgb_to_bgr | x - the tensor to transform | Transform the tensor that was originally in RGB channels to BGR.
+
+### Visualization With TensorBoard
+
+{insert image of tensorboard here}
+
+TensorBoard is a visualization tools provided by Tensorflow to see the graph of your neural network. DIGITS provides easy access to TensorBoard for your network while creating it. The TensorBoard can be accessed by clicking on the ```Visualize``` button under ```Custom Network``` as seen in the image below.
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: tools-> tool

File: docs/GettingStartedTensorflow.md [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+classification_accuracy | pred - the image to be classified <br> y - the labels | Used to measure how accurate the classification task is
+nhwc_to_nchw | x - the tensor to transpose | Transpose the tensor that was originally NHWC format to NCHW. The tensor must be a degree of 4
+nchw_to_nhwc | x - the tensor to transpose | Transpose the tensor that was originally NCHW format to NHWC. The tensor must be a degree of 4
+hwc_to_chw | x - the tensor to transpose | Transpose the tensor that was originally HWC format to CHW. The tensor must be a degree of 3
+chw_to_hwc | x - the tensor to transpose | Transpose the tensor that was originally CHW format to HWC. The tensor must be a degree of 3
+bgr_to_rgb | x - the tensor to transform | Transform the tensor that was originally in BGR channels to RGB.
+rgb_to_bgr | x - the tensor to transform | Transform the tensor that was originally in RGB channels to BGR.
+
+### Visualization With TensorBoard
+
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: Add:
_Experimental functionality._

File: docs/GettingStartedTensorflow.md [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+{insert image of tensorboard here}
+
+TensorBoard is a visualization tools provided by Tensorflow to see the graph of your neural network. DIGITS provides easy access to TensorBoard for your network while creating it. The TensorBoard can be accessed by clicking on the ```Visualize``` button under ```Custom Network``` as seen in the image below.
+
+{insert image here}
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: +

File: docs/GettingStartedTensorflow.md [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+nhwc_to_nchw | x - the tensor to transpose | Transpose the tensor that was originally NHWC format to NCHW. The tensor must be a degree of 4
+nchw_to_nhwc | x - the tensor to transpose | Transpose the tensor that was originally NCHW format to NHWC. The tensor must be a degree of 4
+hwc_to_chw | x - the tensor to transpose | Transpose the tensor that was originally HWC format to CHW. The tensor must be a degree of 3
+chw_to_hwc | x - the tensor to transpose | Transpose the tensor that was originally CHW format to HWC. The tensor must be a degree of 3
+bgr_to_rgb | x - the tensor to transform | Transform the tensor that was originally in BGR channels to RGB.
+rgb_to_bgr | x - the tensor to transform | Transform the tensor that was originally in RGB channels to BGR.
+
+### Visualization With TensorBoard
+
+{insert image of tensorboard here}
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: +

File: docs/GettingStartedTensorflow.md [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+{insert image here}
+
+If there is something wrong with the network model, DIGITS will automatically provide with you the stacktrace and the error message to help you understand where the problem is.
+
+To know more about how TensorBoard works, its official documentation is availabile in the [official tensorflow documentaton](https://www.tensorflow.org/get_started/summaries_and_tensorboard)
+
+## Examples
+
+### Simple Auto-Encoder Network
+
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md:145]: +add one sentence what this does.

File: docs/GettingStartedTensorflow.md (hunk 9 above)
Comment (TimZaman) [reply docs/GettingStartedTensorflow.md]: + add 2 sentences what the below model does (classification of hand written characters to which character it belongs to)

File: docs/GettingStartedTensorflow.md [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+        model = self.inference
+        loss = digits.mse_loss(original, model)
+
+        return loss
+```
+
+### Freezing Variables in Pre-Trained Models by Renaming
+
+The following is a demonstration of how to specifying which weights we would like to use for training. This works best if we are using a pre-trained model. This is applicable for fine tuning a model.
+```python
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: put 5-10 sentences on how this works and how exactly one uses this: train your thing, model is saved, make a new model, point to the saved thing, then change the weight names you dont want, etc.

File: docs/GettingStartedTensorflow.md [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+        return model
+
+    @model_property
+    def loss(self):
+        loss = calculate_loss()
+        """code to calculate loss omitted"""
+        return loss
+```
+
+### Multi-GPU Training
```
Comment (TimZaman) [inline docs/GettingStartedTensorflow.md]: remove

File: docs/BuildTensorflow.md [hunk 21]
```diff
@@ -1,19 +1,44 @@
...
+(https://www.tensorflow.org/versions/master/install/)
+
+Tensorflow comes with pip, to install it, just simply use the command
+```
+pip install tensorflow-gpu=1.1.0
+```
+
+Tensorflow should then install effortlessly and pull in all its required dependices.
+
+## Installation From Source
```
Comment (TimZaman) [inline docs/BuildTensorflow.md]: no no no no

File: docs/BuildTensorflow.md [hunk 22]
```diff
@@ -1,19 +1,44 @@
...
+Tensorflow for DIGITS requires one or more NVIDIA GPUs with CUDA Compute Capbility of 3.0 or higher. See [the official GPU support list](https://developer.nvidia.com/cuda-gpus) to see if your GPU supports it.
+
+Along with that requirement, the following should be installed
+
+* One or more NVIDIA GPUs ([details](InstallCuda.md#gpu))
+* An NVIDIA driver ([details and installation instructions](InstallCuda.md#driver))
+* A CUDA toolkit ([details and installation instructions](InstallCuda.md#cuda-toolkit))
+* cuDNN ([download page](https://developer.nvidia.com/cudnn))
+
+## Basic Installation
```
Comment (TimZaman) [inline docs/BuildTensorflow.md]: ~Basic~

**Review by TimZaman — CHANGES_REQUESTED** [review]
Some changes needed

**Review by TimZaman — APPROVED** [review]
Nice!

**Comment by TimZaman** [comment]
Why does the travis build fail?

**Comment by ethantang95** [comment]
numpy updated to v 1.13 and it broke stuff, it's fixed on master. The other is just linting

**Comment by TimZaman** [comment]
Make sure to keep the CI happy - even the linter.
